operator/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+                    let config = handlebars
+                        .render("conf", &json!({ "options": options }))
+                        .unwrap();
+
+                    let mut data = BTreeMap::new();
+                    data.insert("server.properties".to_string(), config);
+
+                    // And create the ConfigMap
+                    // TODO: Need to deal with the case where the configmaps already exists (this should only happen in unclean shutdowns or similar bad scenarios)
+                    let tmp_name = cm_name.clone() + "-config"; // TODO: Create these names once and pass them around so we are consistent
+                    let cm = stackable_operator::create_config_map(
+                        &self.context.resource,
+                        &tmp_name,
+                        data,
+                    )?;
+                    self.context.client.create(&cm).await?;
+                    pod
+                }
+                Some(pod) => {
+                    trace!(
+                        "Found existing pod [{}] for broker [{}]",
+                        Meta::name(pod),
+                        broker.node_name
+                    );
+                    pod.clone()
+                }
+            };
+
+            // If the pod for this server is currently terminating (this could be for restarts or
+            // upgrades) wait until it's done terminating.
+            if finalizer::has_deletion_stamp(&pod) {
+                info!("Waiting for Pod [{}] to terminate", Meta::name(&pod));
+                return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+            }
+
+            // At the moment we'll wait for all pods to be available and ready before we might enact any changes to existing ones.
+            // TODO: Only do this next check if we want "rolling" functionality
+            if !podutils::is_pod_running_and_ready(&pod) {
+                info!(
+                    "Waiting for Pod [{}] to be running and ready",
+                    Meta::name(&pod)
+                );
+                return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+            }
+        }
+
+        Ok(ReconcileFunctionAction::Continue)
+    }
+
+    pub async fn delete_excess_pods(&self) -> KafkaReconcileResult {
+        trace!("Starting to delete excess pods",);
+
+        // We iterate over all Pods that have an OwnerReference pointing at us
+        // and we'll try to find all pods that are _not_ assigned to any
+        // broker from the spec, we then delete it.
+        for pod in &self.existing_pods {
+            if !self
+                .kafka_spec
+                .brokers
+                .iter()
+                .any(|broker| podutils::is_pod_assigned_to_node(pod, &broker.node_name))
+            {
+                self.context.client.delete(pod).await?;
+            }
+        }
+        Ok(ReconcileFunctionAction::Continue)
```

Comment:
If there were excess pods which we deleted, this was probably a cluster shrink. General question, should we requeue to enable performing checks whether the cluster has rebalanced after the shrink before taking any more actions?
Irrelevant in this case, as it is the last action anyway and there are no checks to perform..

Reply:
Not irrelevant. It's a good question.
Should we delete a single pod and then requeue or should we delete all of them?

In Kafka we probably want to take steps to rebalance before removing too many?
This is where you are the expert.

But before we have a full solution you're right and it might make sense to requeue after every deletion.